### PR TITLE
fix: internally use forward delineated paths cross-platform

### PR DESCRIPTION
**Pull request by mjlbach**
Closes https://github.com/neovim/nvim-lspconfig/issues/1168
Closes https://github.com/neovim/nvim-lspconfig/issues/1170
Closes https://github.com/neovim/nvim-lspconfig/issues/1266
Closes https://github.com/neovim/nvim-lspconfig/issues/1543
Closes https://github.com/neovim/nvim-lspconfig/issues/1588
Closes https://github.com/neovim/nvim-lspconfig/issues/1590

I have not yet tested this.

### CODE REVIEW

File: lua/lspconfig/configs.lua [hunk 1]
```diff
@@ -86,8 +86,9 @@ function configs.__newindex(t, config_name, config_def)
         api.nvim_command(
           string.format(
             "autocmd %s unsilent lua require'lspconfig'[%q].manager.try_add_wrapper()",
-            'BufReadPost ' .. root_dir .. '/*',
-            config.name
+            'BufReadPost %s %s/*',
+            config.name,
+            util.path.sanitize_root_dir(root_dir, true)
```
Comment (kylo252) [inline lua/lspconfig/configs.lua:91]:
```suggestion
            util.path.sanitize_path(root_dir, true)
```
I don't think `sanitize_root_dir` exists

File: lua/lspconfig/util.lua [hunk 2]
```diff
@@ -124,25 +130,41 @@ M.path = (function()
...
     end
   end
 
   local function path_join(...)
-    local result = table.concat(vim.tbl_flatten { ... }, path_sep):gsub(path_sep .. '+', path_sep)
+    local result = table.concat(vim.tbl_flatten { ... }, path_sep()):gsub(path_sep() .. '+', path_sep())
     return result
   end
 
+  local function sanitize_path(path, escape_filename)
```
Comment (kylo252) [inline lua/lspconfig/util.lua:156]:
```suggestion
  local function sanitize(path, escape_filename)
```
nit: I'm thinking it's redundant to call it `util.path.santize_path`.

Is there a reason why we can't always use `fnameescape()`?

File: lua/lspconfig/util.lua [hunk 3]
```diff
@@ -124,25 +130,41 @@ M.path = (function()
...
+          return '/'
+        end
       end
       return result
     end
   end
 
   local function path_join(...)
-    local result = table.concat(vim.tbl_flatten { ... }, path_sep):gsub(path_sep .. '+', path_sep)
+    local result = table.concat(vim.tbl_flatten { ... }, path_sep()):gsub(path_sep() .. '+', path_sep())
```
Comment (kylo252) [inline lua/lspconfig/util.lua:152]:
```suggestion
    local result = table.concat({ ... }, path_sep)
```

**Comment by kylo252** [comment]
`tsserver` is still broken when opening from `netrw`, both on master and on this branch, while opening the file directly works fine on either branches 🤔

**Comment by mjlbach** [comment]
I know the issue, dirname is still failing with the path separators fixed. I'll fix it tomorrow.

lua print(require('lspconfig.util').path.dirname('C:/Users/michael/Repositories/nvim-lspconfig/typescript-project/main.ts'))

vs. 

lua print(require('lspconfig.util').path.dirname('C:\\Users\\michael\\Repositories\\nvim-lspconfig\\typescript-project\\main.ts'))

**Comment by kylo252** [comment]
> I know the issue, dirname is still failing with the path separators fixed because what is returned from netrw is inconsistent. 
> `lua print(require('lspconfig.util').path.dirname('C:\Users\michael\Repositories\nvim-lspconfig\typescript-project\main.ts'))`

I guess you mean with double backslash, because it's throwing an error otherwise

```lua
print(require('lspconfig.util').path.dirname('C:\\Users\\michael\\Repositories\\nvim-lspconfig\\typescript-project\\main.ts'))
```

**Comment by mjlbach** [comment]
I believe this should now work.

**Comment by kylo252** [comment]
> I believe this should now work.

Tested with `tsserver` and it attaches correctly now.

**Comment by sm1999** [comment]
Tested and it fixes #1266 #1588 #1590

**Comment by mjlbach** [comment]
@sm1999 To clarify, does that imply it does *not* fix #1168, #1170, and #1543?

**Comment by sm1999** [comment]
@mjlbach No

**Comment by mjlbach** [comment]
Thanks! I went ahead and merged as I am pretty confident in the fix, but please feel free to report any bugs you all run into.

**Pull request merged**